doc/sr_cpump.1.rst
==========
 SR_CPUMP 
==========

-----------------
sr_shovel in C
-----------------

:Manual section: 1 
:Date: @Date@
:Version: @Version@
:Manual group: MetPX-Sarracenia

.. contents::

SYNOPSIS
========

**sr_cpump** foreground|start|stop|restart|reload|status configfile
**sr_cpump** cleanup|declare|setup configfile

DESCRIPTION
===========

**sr_cpump** is an alternate implementation of `sr_subscribe(7) <sr_subscribe.1.rst>`_ 
with some limitations.  

 - doesn't download data, but only circulates posts.
 - runs as only a single instance (no multiple instances.) 
 - does not support any plugins.
 - does not support vip for high availability.
 - different regular expression library: POSIX vs. python.
 - does not support regex for the strip command (no non-greedy regex)

It can therefore act as a drop-in replacement for:

   `sr_report(1) <sr_report.1.rst>`_ - process report messages.

   `sr_shovel(8) <sr_shovel.8.rst>`_ - process shovel messages.

   `sr_winnow(8) <sr_winnow.8.rst>`_ - process winnow messages.

The C implementation may be easier to make available in specialized environments, 
such as High Performance Computing, as it has far fewer dependencies than the python version.
It also uses far less memory for a given role.  Normally the python version 
is recommended, but there are some cases where use of the C implementation is sensible.

**sr_cpump** connects to a *broker* (often the same as the posting broker)
and subscribes to the notifications of interest. On reception of a post,
it looks up its **sum** in its cache.  if it is found, the file has already come through,
so the notification is ignored. If not, then the file is new, and the **sum** is added 
to the cache and the notification is posted.  

**sr_cpump** can be used, like `sr_winnow(8) <sr_winnow.8.rst>`_,  to trim messages 
from `sr_post(1) <sr_post.1.rst>`_, `sr_poll(1) <sr_poll.1.html>`_  
or `sr_watch(1) <sr_watch.1.rst>`_  etc... It is used when there are multiple 
sources of the same data, so that clients only download the source data once, from 
the first source that posted it.

The **sr_cpump** command takes two argument: an action start|stop|restart|reload|status... (self described)
followed by a configuration file.

The **foreground** action is used when debugging a configuration, when the user wants to 
run the program and its configfile interactively...   The **foreground** instance 
is not concerned by other actions.  The user would stop using the **foreground** instance 
by simply <ctrl-c> on linux or use other means send SIGINT or SIGTERM to the process.

The actions **cleanup**, **declare**, **setup** can be used to manage resources on
the rabbitmq server. The resources are either queues or exchanges. **declare** creates
the resources. **setup** creates and additionally does the bindings of queues.

The actions **add**, **remove**, **edit**, **list**, **enable**, **disable** act
on configurations.

CONFIGURATION
=============

In general, the options for this component are described by 
the `sr_subscribe(1) <sr_subscribe.1.rst>`_  page which should be read first.
It fully explains the option configuration language, and how to find
the option settings.

**NOTE**: The regular expression library used in the C implementation is the POSIX
one, and the grammar is slightly different from the python implementation.  Some
adjustments may be needed.


ENVIRONMENT VARIABLES
=====================

if the SR_CONFIG_EXAMPLES variable is set, then the *add* directive can be used
to copy examples into the user's directory for use and/or customization.

An entry in the ~/.config/sarra/default.conf (created via sr_subscribe edit default.conf )
could be used to set the variable::

  declare env SR_CONFIG_EXAMPLES=/usr/lib/python3/dist-packages/sarra/examples

the value should be available from the output of a list command from the python
implementation.

SEE ALSO
========

`sr_report(7) <sr_report.7.rst>`_ - the format of report messages.

`sr_report(1) <sr_report.1.rst>`_ - process report messages.

`sr_post(1) <sr_post.1.rst>`_ - post announcemensts of specific files.

`sr_post(7) <sr_post.7.rst>`_ - The format of announcements.

`sr_subscribe(1) <sr_subscribe.1.rst>`_ - the download client.

`sr_watch(1) <sr_watch.1.rst>`_ - the directory watching daemon.

`dd_subscribe(1) <dd_subscribe.1.rst>`_ - the http-only download client.


CHANGES.rst
===============
 Release Notes 
===============

lists all changes between versions.

**2.18.06b1**

* new         start post (sleep <= 0) now does nothing.
* doc         French docs done.
*             *sci-fi* (future planned features) removed from docs.
*             website migrated from sf.net to git repo ones (github, gitlab)

**2.18.05b4**

* bugfix      plugin msg_filter_wmo2msc.py fixed
*             sr_poll now uses its own set_dir_pattern to replace variables
*             sr_post/sr_poll/sr_watch fixed that a cache file for instance 000 was created
*             sr_post/sr_poll/sr_watch add caching of 'L' message
* new         sr_amqp for consumer sets channel.basic_recover(requeue=True)
* update      flow_post.sh posts lot of files within one call (load problem fixed)
*             flow_test does not test filename with space anymore
*             flow_test 5 instances for sender
* doc         miscellaneous doc fixups.

**2.18.05b2**

* bugfix       sr_sftp use chdir to see if connection is still alive
*              sr_http some site do not tolerate '//' in path
*              sr_subscribe some code added for sr_sender (lost in inheritance)
*              sr_file if a file does not exist and should be copied/linked just warn (Dominic)
* update       flow_test test2_f61.conf no use of post_total_save (race condition)

**2.18.05b1**

* bugfix       sr_sftp  differentiate put part vs put file 
* new          sr_retry uses caching for message uniqueness
*              sr_ftp,sr_sftp better connection test
* update       sr_poll  default post_base_url from destination url without password
*              sr_subscribe,sr_sender  log attempt
*              sr_subscribe retry.on_heartbeat on startup
*              sr_sender/sr_util if file to send does not exist... give it "attempt" chances
* doc          fixes and translation into french in progress
*              updated examples link, and samples directory cleanup

**2.18.04b5**

* bugfix  open local file read rb (instead of r+b)
*         remove .tmp file if upload does not work
*         sftp link and directory removal (from message)
* update  instance string using 2 digits (and code to migrate to)
*         switch for subprocess.check_call or subprocess.run 
*         give all chances to sigterm to complete
* new     inflight tmp/
*         If things go badly (general exception catch) keep message in retry
*         log when a retry message is dropped because it expires
*         poll accepts https (more test to do...not working with japan)
* doc     expanded description of *expire* setting.
*         Adding hint about use of *expire* option, how it is necessary for operations.
*         default setting is to avoid broker overload, too low for operational use.

**2.18.04b4**

* bugfix  destfn_script was not working
* update  sr_config.run_command presents subprocess check_call or run depending on python version
* new     inflight tmp/

**2.18.04b3**

* testing flow_test standardisation of messages
* bugfix  hb_retry no more dependency with on_start
*         sr_poll make sure all comparaisons are done without of trailing \n
* new     registered do_get,do_put
*         registered plugin, if return None, let python do it
*         plugin accel_scp.py
*         C has its own sarrac git repo
*         sr_config.py  '\ ' backslash_space allowed in options 
*         sr_config.py/sr_message.py  topic encodes  ' ' '#' with %20 %23
* update  flow_test standardisation of messages
*         sr_poll logs a warning when sleep time makes no sense
*         documentation launchpat for sundew
*         sr_audit log a message when --users all is done
*                   make sure heartbeat is in try/except 

**2.18.04b2**

* testing: Added recovery of flow_test stuck in retrying state.
* bugfix  C Truncate all headers and topic so they don't exceed 255 (AMQP limit.)
*         C Try to avoid being in conflict with stdin/out/err  open/close + 2 dup
*         C Valgrind hygiene: if nanosecond timestamp was 0, weird stuff happenned.  Fixed.
*           now valgrind does not complain at all.
*         C libsrshim enforced checks on commands'status
*         C any Python, topic and path with # encoded into %23 (as blank into %20)
* update  sr_audit hb_police_queues to check queue as admin, 
*           hb_sanity to check processes and sanity_log_dead option added            
*           no sleep option, sleep computed to trigger next heartbeat
*         sr_rabbit rabbit dependant commands placed in this file
* new     plugin do_send_log

**2.18.04b1**

* testing:retries on python side to validate products and routing
*         flow_post: loop on sorted individial products (spaces in path)... symlinks considered
* new     sr_subscribe: traceback logs when doing badly
*         sr_audit heartbeat works ... needs a config in audit/x.conf for now.
*         plugins: hp_sanity  uses sr sanity to check if program in strange state and log age to restart
*                  do_simulation logs protocol steps... instead of doing the actual download or send of a product
*         sr_config sundew_dirpattern provide a mean to use $RYYYY... etc in directory
* bugfix  C changes to return proper status of shimmed functions
*         C Get log file descriptor out of the danger zone also
*         C renameorlink put back code when oldname exists and processes it if link too
*         sr_retry : no more uses of self.activity and conditional retry heartbeat changed

**2.18.03b1**

* testing: changes derived expanded flow_test coverage
*        plugin msg_stopper with env MAX_MESSAGES
*        filename with spaces: ls_file_index (poll,sftp,ftp), sr_post.c, flow_post.sh
*        flow config changes : reject (hourly,today,yesterday xml)
*        plugins   : msg_pclean_f9*.py
*        sr_subscribe logging fix
* new    realpath_filter (PY and C), realpath also named realpath_post
*        sr sanity  check pid/process  and log age if older than heartbeat * 1.5

c/sr_cpump.c
/* vim:set ft=c ts=2 sw=2 sts=2 et cindent: */

/*
 * Usage info after license block.
 *
 * This code is by Peter Silva copyright (c) 2017 part of MetPX.
 * copyright is to the Government of Canada. code is GPL.
 *
 */

/* 
  Minimal c implementation to allow posting of sr_post(7) messages.

  spits out ´unimplemented option´ where appropriate...
 */
#include <stdio.h>
#include <signal.h>
#include <unistd.h>

//for opendir/readdir
#include <dirent.h>



#include "sr_version.h"

#include "sr_consume.h"
#include "sr_post.h"

void usage()
{
    fprintf( stderr, "usage: sr_cpump %s <options> <action> <configuration>\n\n", __sarra_version__  );
    fprintf( stderr, "\t<options> - sr_post compatible configuration file.\n" );
    fprintf( stderr, "\tbroker amqps://<user>@host - required - to lookup in ~/.config/sarra/credentials.\n" );
    fprintf( stderr, "\tdebug <on|off> - more verbose output.\n" );
    fprintf( stderr, "\texchange <exchange> - required - name of exchange to bind to\n" );
    fprintf( stderr, "\taccept/reject <regex> - to filter files to post.\n" );
    fprintf( stderr, "\tqueue <name> - queue to declare & use.\n" );

    fprintf( stderr, "\tpost_exchange <exchange> name of exchange to publish to (default: xs_<brokerusername>.)\n" );
    fprintf( stderr, "\tpost_exchange_split <num> number of output exchanges (default: 1.)\n" );
    fprintf( stderr, "\theartbeat <on|off|integer> - clean cache interval.\n" );
    fprintf( stderr, "\tsuppress_duplicates|sd|cache|caching <on|off|integer> (default: off)\n" );
    fprintf( stderr, "\t\tsuppress duplicate receptions < *cache* seconds apart.  \"on\" means 15 minute caching (on=900).\n" );
    fprintf( stderr, "\t\tsuppress duplicate announcements < *cache* seconds apart.  \"on\" means 15 minute caching (on=900).\n" );
    fprintf( stderr, "\ttopic_prefix <string> - AMQP topic prefix (default: v02.post )\n" );
    fprintf( stderr, "\tto <destination> - clusters pump network should forward to (default: broker).\n" );
    fprintf( stderr, "\turl <url>[,<url>]... - retrieval base url in the posted files.\n" );
    fprintf( stderr, "\n\t<action> = [start|stop|setup|cleanup|foreground] default: foreground\n" );
    fprintf( stderr, "\t\tstart - start a daemon running (will detach) and write to log.\n" );
    fprintf( stderr, "\t\thelp - view this usage.\n" );
    fprintf( stderr, "\t\tlist - list existing configurations.\n" );
    fprintf( stderr, "\t\tstop - stop a running daemon.\n" );
    fprintf( stderr, "\t\tdeclare - declare broker resources (to be ready for subscribers to bind to.)\n" );
    fprintf( stderr, "\t\tsetup - bind queues to resources, declaring if needed.)\n" );
    fprintf( stderr, "\t\tcleanup - delete any declared broker resources.\n" );
    fprintf( stderr, "\t\tforeground - run as a foreground process logging to stderr (ideal for debugging.)\n" );
   
    fprintf( stderr, "\nThis is a limited C implementation of sr_subscribe(1), see man page for details\n\n" );
    fprintf( stderr, "\t\tcan only move messages (no downloading.), no plugin support.\n" );
    exit(1);
}

int sr_cpump_cleanup(struct sr_context *sr_c, struct sr_config_t *sr_cfg, int dolog)
{
  DIR   *dir;
  int    ret;
  char   cache_dir[PATH_MAX];
  char   cache_fil[PATH_MAX];
  struct stat sb;
  struct dirent *e;

  // if running, warn no cleanup
  if (sr_cfg->pid > 0)
  {
     ret=kill(sr_cfg->pid,0);
     if (!ret)
     {   // is running.
         fprintf( stderr, "cannot cleanup : sr_cpump configuration %s is running\n", sr_cfg->configname );
         return(1);
     }
  }

  sprintf( cache_dir, "%s/.cache/sarra/%s/%s", getenv("HOME"), sr_c->cfg->progname, sr_c->cfg->configname);

  sr_consume_cleanup(sr_c);
  sr_post_cleanup( sr_c ); 
  sr_context_close(sr_c);

  dir = opendir( cache_dir );

  if (dir)
  {
      while( (e = readdir(dir)) )
      {
          if ( !strcmp(e->d_name,".") || !strcmp(e->d_name,"..") )
               continue;

          strcpy( cache_fil, cache_dir );
          strcat( cache_fil, "/" );
          strcat( cache_fil, e->d_name );

          if ( lstat( cache_fil, &sb ) < 0 )
               continue;

          if ( S_ISDIR(sb.st_mode) )
          {
               fprintf( stderr, "cannot cleanup : sr_cpump configuration %s directory\n", e->d_name );
          }

          ret = remove(cache_fil);
      }

      closedir(dir);

      ret = rmdir(cache_dir);
  }

  /* PAS not sure, but currently don't think we should ever delete logs.
     MG also mentioned this code does not delete old logs, so missing a bit.
  if (dolog)
  {
     ret = remove(sr_cfg->logfn);
  }
   */
  return(0);
}

int main(int argc, char **argv)
{
  struct sr_message_t *m;
  struct sr_context *sr_c;
  struct sr_config_t sr_cfg;
  struct sr_mask_t *mask;
  int    consume,i,ret;
  char   *one;
  
  //if ( argc < 3 ) usage();
 
  sr_config_init( &sr_cfg, "cpump" );

  i=1;
  while( i < argc ) 
  {
      if (argv[i][0] == '-') 
         consume = sr_config_parse_option( &sr_cfg, 
              &(argv[i][ (argv[i][1] == '-' )?2:1 ]),  /* skip second hyphen if necessary */
              argv[i+1], 
              (argc>i+2)?argv[i+2]:NULL, 
              1 );
      else
          break;
      if (!consume) break;
      i+=consume;
  }

  for (; i < argc; i++ )
  {
        sr_add_path(&sr_cfg, argv[i]);
  }

  if ( !strcmp( sr_cfg.action, "add" ))
  {
        sr_config_add( &sr_cfg );
        exit(0);
  }
  
  if ( !strcmp( sr_cfg.action, "enable" ))
  {
        sr_config_enable( &sr_cfg );
        exit(0);
  }
  
  if ( !strcmp( sr_cfg.action, "disable" ))
  {
        sr_config_disable( &sr_cfg );
        exit(0);
  }

  if ( sr_cfg.paths )
  {
        sr_config_read(&sr_cfg, sr_cfg.paths->path, 1, 1 );
  }

  if ( sr_cfg.help ) usage();

  if ( !strcmp( sr_cfg.action, "edit" ))
  {
        sr_config_edit( &sr_cfg );
        exit(0);
  }

  if ( !strcmp( sr_cfg.action, "remove") )
  {
      // remove anything but a config file
      if (sr_cfg.configname)
      {
         one = sr_config_find_one( &sr_cfg, sr_cfg.configname );
         if ( !one || strcmp( &(one[strlen(one)-5]),".conf"))
         {
            sr_config_remove( &sr_cfg );
            exit(0);
         }
      }
  }

  if ( !strcmp( sr_cfg.action, "list" ))
  {
        sr_config_list( &sr_cfg );
        exit(0);
  }
  
  if (!sr_config_finalize( &sr_cfg, 1))
  {
     log_msg( LOG_ERROR, "failed to finalize configuration\n");
     return(1); 
  }

  if ( !strcmp( sr_cfg.action, "log" ))
  {
        sr_config_log( &sr_cfg );
        exit(0);
  }

    // Check if already running. (conflict in use of state files.)

    ret = sr_config_startstop( &sr_cfg );
    if ( ret < 1 )
    {
        exit(abs(ret));
    }

  sr_c = sr_context_init_config( &sr_cfg );
  if (!sr_c) {
     log_msg( LOG_ERROR, "failed to build context from configuration\n");
     return(1);
  }
  sr_c = sr_context_connect( sr_c );
  if (!sr_c) {
     log_msg( LOG_ERROR, "failed to connect context.\n");
     return(1);
  }
  
  if ( !strcmp( sr_cfg.action, "remove" ))
  {

        ret = sr_cpump_cleanup(sr_c,&sr_cfg,1);
        if (ret == 0) sr_config_remove( &sr_cfg );
        exit(0);
  }

  // dont consume_setup or post_init if in cleanup
  // (just hangs when attempting to bind queue with cleaned up exchange)
  if ( strcmp( sr_cfg.action, "cleanup" ) ) {
     sr_consume_setup(sr_c);

     if (!strcmp(sr_cfg.outlet,"post"))
         sr_post_init( sr_c );
  }
  
  if ( !strcmp( sr_cfg.action, "setup" ) || !strcmp( sr_cfg.action, "declare") )
  {
        return(0);
  }

  if ( !strcmp( sr_cfg.action, "cleanup" ) )
  {
        ret = sr_cpump_cleanup(sr_c,&sr_cfg,0);
        return(0);
  }

  /*
  if ( !strcmp( sr_cfg.action, "start" ) )
  {
      log_msg( LOG_CRITICAL, "FIXME output file when running as daemon is broken, aborting.\n");
      return(0);
  }
   */

  if ( strcmp( sr_cfg.action, "foreground" ) )
  {
      if (! sr_cfg.outlet) 
      {
           log_msg( LOG_CRITICAL, "must specify output file when running as daemon.\n");
           return(1);
      }
      daemonize(0);
  }
 
  // Assert: this is a working instance, not a launcher...
  if ( sr_config_activate( &sr_cfg ) )
  {
        log_msg( LOG_WARNING, 
            "could not save pidfile %s: possible to run conflicting instances  \n", 
            sr_cfg.pidfile );
  }
  log_msg( LOG_INFO, "%s %s config: %s, pid: %d, queue: %s bound to exchange: %s starting\n", 
            sr_cfg.progname, __sarra_version__, sr_cfg.configname,  sr_cfg.pid, sr_cfg.queuename, sr_cfg.exchange );


  while(1)
  {

      // inlet: from queue, json, tree.
      m=sr_consume(sr_c);

      if (!m) break;

      log_msg( LOG_INFO, "received: %s\n", sr_message_2log(m) );

      /* apply the accept/reject clauses */
      // FIXME BUG: pattern to match is supposed to be complete URL, not just path...
      mask = isMatchingPattern( &sr_cfg, m->path);
      if ( (mask && !(mask->accepting)) || (!mask && !(sr_cfg.accept_unmatched)) )
      {
          log_msg( LOG_DEBUG, "rejecting: %s\n", m->path );
          continue; 
      }
      // check cache.
      if (sr_cfg.cachep)
      {
          ret = sr_cache_check( sr_cfg.cachep, m->parts_s, (unsigned char*)m->sum, m->path, sr_message_partstr(m) );
          if (!ret) continue; // cache hit.
      }
      // do_pump=NULL
 
      // outlet:
      if (m) {
        if ( !strcmp( sr_cfg.outlet, "json" ) ) sr_message_2json(m);      
        else if ( !strcmp( sr_cfg.outlet, "url" ) ) sr_message_2url(m);      
        else if ( !strcmp( sr_cfg.outlet, "post" ) ) sr_post_message(sr_c,m);      
      }

      sr_context_heartbeat_check( sr_c );

  }
  sr_context_close(sr_c);
  free(sr_c);
  sr_config_free(&sr_cfg);

  return(0);
}
